inc/packages/admin/namespace.php

```diff
@@ -0,0 +1,312 @@
+}
+
+/**
+ * Filters the tabs shown on the Add Plugins screen.
+ *
+ * @param string[] $tabs Map of tab ID to map name.
+ * @return string[] Updated tabs.
+ */
+function add_direct_tab( $tabs ) {
+	$tabs[ TAB_DIRECT ] = __( 'Direct Install', 'fair' );
+	return $tabs;
+}
+
+/**
+ * Handles the AJAX request for plugin information when a DID is present.
+ *
+ * @param mixed  $result The result of the plugins_api call.
+ * @param string $action The action being performed.
+ * @param object $args   The arguments passed to the plugins_api call.
+ * @return mixed
+ */
+function handle_did_during_ajax( $result, $action, $args ) {
+	if ( ! wp_doing_ajax() || 'plugin_information' !== $action || ! isset( $args->slug ) ) {
+		return $result;
+	}
+
+	$slug = sanitize_text_field( $args->slug );
+	if ( ! str_contains( $slug, '-did--' ) ) {
+		return $result;
+	}
+
+	$did = 'did:' . explode( '-did:', str_replace( '--', ':', $slug ), 2 )[1];
+	if ( ! preg_match( '/^did:(web|plc):.+$/', $did ) ) {
+		return $result;
+	}
+
+	wp_cache_set( ACTION_INSTALL_DID, $did );
+	Updater\add_package_to_release_cache( $did );
+	add_filter( 'http_request_args', 'FAIR\\Updater\\maybe_add_accept_header', 20, 2 );
+
+	return (object) Packages\get_update_data( $did );
+}
+
+/**
+ * Enqueue assets.
+ *
+ * @return void
+ */
+function load_plugin_install() {
+	enqueue_assets();
+}
+
+/**
+ * Enqueue assets.
+ *
+ * @return void
+ */
+function enqueue_assets() {
+	wp_enqueue_style(
+		'fair-admin',
+		esc_url( plugin_dir_url( FAIR\PLUGIN_FILE ) . 'assets/css/packages.css' ),
+		[],
+		FAIR\VERSION
+	);
+}
+
+/**
+ * Render direct installer tab.
+ *
+ * @return void
+ */
+function render_tab_direct() {
+	?>
+	<div class="fair-direct-install">
+		<p class="fair-direct-install__help">
+			<?= __( "If you have a plugin's ID, enter it here to view the details<br />and install it.", 'fair' ); ?>
+		</p>
+		<form
+			action=""
+			class="fair-direct-install__form"
+			method="post"
+		>
+			<input
+				type="hidden"
+				name="tab"
+				value="<?= esc_attr( TAB_DIRECT ); ?>"
+			/>
+			<?php wp_nonce_field( TAB_DIRECT ); ?>
+			<label
+				class="screen-reader-text"
+				for="plugin_id"
+			>Plugin ID</label>
+			<input
+				type="text"
+				id="plugin_id"
+				name="plugin_id"
+				pattern="did:(web|plc):.+"
+				placeholder="did:..."
+				aria-describedby="fair-direct-install__note"
+			/>
+			<?php submit_button( _x( 'View Details', 'plugin', 'fair' ), '', '', false ); ?>
+		</form>
+		<p id="fair-direct-install__note">
+			<?= __( 'Plugin IDs should be in the format <code>did:web:...</code> or <code>did:plc:...</code>', 'fair' ); ?>
+		</p>
+	</div>
+	<script>
+		// On submit, trigger the thickbox with the plugin information.
+		document.querySelector( '.fair-direct-install__form' ).addEventListener( 'submit', function ( e ) {
+			e.preventDefault();
+			e.stopPropagation();
+			const id = document.getElementById( 'plugin_id' ).value;
+
```

Comment:
I'd get this constant outside of the event listener so it can be used as a conditional to prevent execution or trigger an error message when there's no DID added.

Reply:
Making the input field required makes this largely irrelevant. If we later add some kind of pre-submission validation, it could be useful, but at the moment I don't think it's relevant.